swarmform/core/cluster.py

```diff
@@ -0,0 +1,180 @@
+    for fw_id in fw_ids:
+        if fw_id not in parallel_fw_ids:
+            firetask_list = swarmpad.get_fw_by_id(fw_id).spec['_tasks']
+            num_firetasks = len(firetask_list)
+            # Check whether a firework has no firetasks
+            if num_firetasks == 0:
+                raise ValueError('No Firetasks available in the Firework')
+            # Check whether the firework has multiple firetasks - Not checking firetasks recursively at the moment
+            if num_firetasks > 1:
+                # If multiple firetasks are available, add each task to firetask list
+                for firetask_index in range(0, num_firetasks):
+                    firetask.append(firetask_list[firetask_index])
+            # If only a single firetask is there, get the first firetask and add it to the combined firetask
+            else:
+                firetask.append(firetask_list[0])
+
+    # Create a firework from the combined firetasks
+    combined_fw = Firework(firetask)
+    swarmpad.m_logger.info('Sequentially clustered {} Fireworks to firework_id {}'.format(fw_ids, combined_fw.fw_id))
+    return combined_fw
+
+
+# TODO: Resolve the following assumptions
+'''
+* Only scriptTasks are given
+* Command in scriptTasks are static
+* fws to combine parallely have only a single firetask
+'''
+
+
+def combine_fws_parallely(swarmpad, fw_ids):
+    """
+    Combine a set of firetasks into a single firetask which runs all the given tasks parallely
+
+    Args:
+        swarmpad (SwarmPad)
+        fw_ids (list): id of the fireworks to be combined
+
+    Returns:
+        combined_firework (FireWork): Parallely combined FireWork object
+    """
+
+    # Get each task in each firework and append to firetask list in the order of traversal
+    firetasks_to_combine = []
+    for fw_id in fw_ids:
+        firetask_in_fw = swarmpad.get_fw_by_id(fw_id).spec['_tasks'][0]
+        if isinstance(firetask_in_fw, ScriptTask):
+            firetasks_to_combine.append(firetask_in_fw)
+        else:
+            raise ValueError('Spec of Firework with id {} does not contain an object of type ScriptTask '.format(fw_id))
+
+    combined_firetask = ParallelTask.from_firetasks(firetasks_to_combine)
+    combined_firework = Firework(combined_firetask)
+    swarmpad.m_logger.info('Parallely Clustered {} to firework_id {}'.format(fw_ids, combined_firework.fw_id))
+
+    return combined_firework
+
+
+def update_parent_child_relationships(links_dict, old_id, new_id):
+    """
+        Update the parent-child relationships after clustering a firework
+        by replacing all the instances of old_id with new_id
+        Args:
+            links_dict (list): Existing parent-child relationship list
+            old_id (int): Existing id of the firework
+            new_id (int): New id of the firework
+
+        Returns:
+            links_dict (list): Updated parent-child relationship list
+        """
```

Comment:
wrong indentation level for comment and closing """

Reply:
Fixed